python39Packages.opensfm: init at 0.5.2

SomeoneSerge opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Adds OpenSfM - an opensource structure from motion (reconstruction of a sparse 3d point cloud from images, similar to `nixpkgs#colmap`) framework maintained by Mapillary

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

SomeoneSerge committed:
python3Packages.opensfm: init at 0.5.2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/672

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,106 @@
+        description = "Open source Structure-from-Motion pipeline";
+      };
+    };
+in
```
I don't think we need this let in construct for everything. It over complicated things a bit.

SomeoneSerge replied:
Yes, I'll remove it. I was playing with splitting the binary part of the package before submission and left out the `let ... in`

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,106 @@
+        "-Bcmake_build"
+        "-Sopensfm/src"
+      ];
+      patches = [ ./fix-cmake.patch ];
```
Can we fetch this from upstream? It's there a chance we could upstream this?

SomeoneSerge replied:
This cannot be fetched from upstream. The upstream is using custom `Find*.cmake` modules and distributes an intree copy of `gtest` which breaks automatic cmake package discovery. I'm deleting these `Find*` files and the gtest copy, adjusting the root `CMakeLists.txt` accordingly

We cannot upstream this right away, because it's a lot of changes I might struggle to defend, but I'll definitely try to get in touch with them

SomeoneSerge replied:
@SuperSandro2000 may I mark these two comments as resolved?

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,106 @@
+      ];
+      outputs = [ "out" "viewer" ];
+      postInstall = ''
+        cp --no-preserve=mode -r ./viewer/ $viewer
```
This should use moveToOutput.

SomeoneSerge replied:
The change for c++ standard from c++11 to c++14 is already there, but fixing Eigen, etc - not, and I suspect it'd take a lot of time to get it there

I definitely will bring it up, but I'd rather start with the current approach

SomeoneSerge replied:
@SuperSandro2000 I'm copying this path from the source directory, not from `$out`. I've tried `moveToOutput` and it only seems to handle paths in other outputs.

By the way, there seems to be no error handling in that function. When it didn't find any matches for the `viewer/` pattern it silently proceeded with the build and the first error I got was that the `$viewer` output was expected but hasn't been created by the build

SomeoneSerge replied:
On the second thought, the derivation's user can always ask for `${opensfm.src}/viewer`, so I'm removing the $viewer output to keep things simple

SomeoneSerge commented:
I've found this suffers from spurious segfaults when using HAHOG descriptors or when using multithreading or just randomly when computing depthmaps, so I'm converting this into a draft until I'm done with debugging

FRidh commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,111 @@
+}:
+
+let
+  glog' = glog.overrideAttrs (attrs: attrs // {
```
`attrs //` can be removed because overriding is already doing set merging

SomeoneSerge replied:
Thanks! I will remove the whole glog thing, I was just testing a hypothesis...

SomeoneSerge commented:
No more disabled tests, and no segfaults!

I've bumped this to the (unreleased) `0.5.2` because the last release appeared to have some cxx standard issues that are only pronounced when using the new ceres (`std::integer_sequence<>`...), hotfixing which would've left a bunch of runtime errors.

**I think the derivation is ready to be merged.** Now that everything works I would reach out the upstream developers: hopefully they'd be willing to cooperate and remove the need for all the patches I've introduced. If that works out, I'll open a follow-up PR to remove the ad hoc hacks and ensure integration with `nixpkgs-update`

---

One more thing, I've been checking shebangs manually as follows:

```
nix build .#python3Packages.opensfm
./result/bin/opensfm_run_all ${opensfm-src}/data/berlin # modifies data/berlin in-place
```

(there's been a lot of issues with shebangs)

How should I automate this sort of a test?

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,110 @@
+  meta = {
+    maintainers = [ lib.maintainers.SomeoneSerge ];
+    license = lib.licenses.bsd2;
+    description = "Open source Structure-from-Motion pipeline from Mapillary";
```
missing homepage and I suspect that we might need to set platform to linux but that needs to be rechecked.

SomeoneSerge replied:
I've set the homepage and I'll try to find means to test the build under darwin - I think it should still work there too. If it does, I'll assign that and `linux-x86_64`

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,110 @@
+    "-Sopensfm/src"
+  ];
+  patches = [
+    ./fix-cmake.patch
+    ./fix-scripts.patch
+    ./fix-report-generation.patch
```
It would be great if we could try to upstream the none NixOS specific changes.

SomeoneSerge replied:
You're right, I should get `fix-report-generation.patch` merged upstream before this PR can get a go (I've opened a quick [PR](https://github.com/mapillary/OpenSfM/pull/854)), but the other two
- make distribution-related adjustments (like fixing shebangs and pointing cmake at different places)
- are probably bit harder to get approved upstream

SomeoneSerge replied:
But in their CI they seem to have passing tests, it's quite likely nixpkgs just packages a different release of `fpdf` 🤔

SuperSandro2000 replied:
You can also fetch the patch from the PR. Than everyone knows it is already being upstreamed.

We have the latest and greated fpdf version https://repology.org/project/python:fpdf/versions

SomeoneSerge replied:
> You can also fetch the patch from the PR. 

Sounds reasonable. Done

> We have the latest

I thought to check if they use an old one, but in fact they use an alternative package, `fpdf2` =\ Maybe will add it tomorrow

SomeoneSerge committed:
python3Packages.opensfm: fetch patches from PRs

SomeoneSerge committed:
python3Packages.opensfm: limit platforms to linux

SomeoneSerge commented:
Checked this on darwin: it builds, but the tests crash:

```
 python3.8-OpenSfM> collecting ... objc[4817]: Class CaptureDelegate is implemented in both /nix/store/w9gqadll6frp6sw913hlq5r0cylhzrqm-opencv-4.5.2/lib/libopencv_videoio.4.5.dylib (0x108d92630) and /nix/store/iffy67h7fw84ba9n25nw68is7wzhc4nr-opencv-4.5.2/lib/libopencv_videoio.4.5.dylib (0x129ef9630). One of the two will be used. Which one is undefined.
...
python3.8-OpenSfM> [libprotobuf ERROR google/protobuf/descriptor_database.cc:641] File already exists in database: opencv-caffe.proto
python3.8-OpenSfM> [libprotobuf FATAL google/protobuf/descriptor.cc:1371] CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size):
python3.8-OpenSfM> libc++abi.dylib: terminating with uncaught exception of type google::protobuf::FatalException: CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size):
python3.8-OpenSfM> Fatal Python error: Aborted
```

Does anyone have an idea what's this about?

Full logs: https://github.com/SomeoneSerge/pkgs/runs/4880547540?check_suite_focus=true#step:6:2419

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+, fpdf
+}:
+
+let
+  pendingPRs.fixReports = fetchurl {
+    url = "https://github.com/mapillary/OpenSfM/commit/7c4477c6416417a935dd8cfb59e6e7e60cc2ba37.patch";
+    sha256 = "sha256-i6p/K2vIKouY8Fmp9iKdc7Xk2u7o51PQT5fhnKzqmhs=";
+  };
+in buildPythonPackage rec {
```
```suggestion
buildPythonPackage rec {
```

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+  patches = [
+    ./fix-cmake.patch
+    ./fix-scripts.patch
+    (toString pendingPRs.fixReports)
```
```suggestion
    (fetchpatch {
      url = "https://github.com/mapillary/OpenSfM/commit/7c4477c6416417a935dd8cfb59e6e7e60cc2ba37.patch";
      sha256 = "sha256-....";
    })
```
Please update the hash.

SomeoneSerge replied:
the `fpdf` thing got merged upstream

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+    # Fixing by commit-id until the upstream issues are solved
+    rev = "2a1c9be07f47f6e26f87fedf8275fcf8ff7b8487";
+    sha256 = "sha256-CARPFuYypgV1YcqZp//Qws8YrV25Z+w/dWB7MbFAKys=";
+    fetchSubmodules = true;
```
Why do we use their pybind11? Can't we use the nixpkgs one?

SomeoneSerge replied:
I added the changes necessary to use nixpkgs' pybind, but I'm considering reverting them: they used an older version of pybind11 and had their own versions of utilities that have been later introduced upstream (ie in the new pybind11). These are inconsistent

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+  };
+in buildPythonPackage rec {
+  pname = "OpenSfM";
+  version = "0.5.2";
+  src = fetchFromGitHub {
```
```suggestion
  version = "0.5.2";

  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+    "-Bcmake_build"
+    "-Sopensfm/src"
+  ];
+  patches = [
```
patches and postPatch should come after src.

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,118 @@
+    # where segfaults might be introduced in future
+    echo 'feature_type: SIFT' >> data/berlin/config.yaml
+    echo 'feature_type: HAHOG' >> data/lund/config.yaml
+  '';
+  pythonImportsCheck = [ "opensfm" ];
+  meta = {
```
```suggestion
  '';

  pythonImportsCheck = [ "opensfm" ];

  meta = {
```

SomeoneSerge committed:
opensfm: fpdf patch merged, bump revision

SomeoneSerge committed:
opensfm: move src and patches closer together

SomeoneSerge committed:
opensfm: use the nixpkgs-packaged pybind

This allows not to fetch the submodules.
Additionally, the original eigen cmake patch missed out ceres
(miraculously it worked for some opensfm revisions, and gave unresolved
symbols on others, at runtime).

SomeoneSerge committed:
opensfm: run alejandra

SomeoneSerge committed:
opensfm: ceres usage compatible with old nixpkgs

In 21.05 ceres hasn't been split yet

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -1,39 +1,39 @@
-, sphinx
-, matplotlib
-, fpdf
+{
+  lib,
+  buildPythonPackage,
```
Please revert to the previous formatting.

SomeoneSerge replied:
Yeah, I thought the same

SomeoneSerge committed:
opensfm: back to nixpkgs-fmt

SomeoneSerge committed:
opensfm: eigen cmake target patch from PR

SomeoneSerge committed:
opensfm: fix for darwin

- avoid having two opencvs in closure
- disable tests with lowerbounded residuals (smh darwin build converges
  to better minima - maybe different fp precision?)

SomeoneSerge committed:
opensfm: disable tests only for darwin

SomeoneSerge commented:
- [x] darwin runtime errors were caused by having two builds of opencv in the closure - fixed
- [x] some tests had to be disabled for darwin until https://github.com/mapillary/OpenSfM/issues/874 has been fixed

SomeoneSerge commented:
Actually, is it at all a good idea to have something as big and brittle as opencv4 in `propagatedBuildInputs`?

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -112,13 +110,18 @@ buildPythonPackage rec {
     description = "Open source Structure-from-Motion pipeline from Mapillary";
     homepage = "https://opensfm.org/";
-    platforms = lib.platforms.linux;
+    platforms = lib.platforms.linux ++ lib.platforms.darwin;
```
```suggestion
    platforms = lib.platforms.unix;
```

SomeoneSerge committed:
opensfm: refactor platforms linux+darwin=unix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on pkgs/development/python-modules/opensfm/default.nix:
```diff
@@ -0,0 +1,128 @@
+    license = lib.licenses.bsd2;
+    description = "Open source Structure-from-Motion pipeline from Mapillary";
+    homepage = "https://opensfm.org/";
+    platforms = lib.platforms.unix;
```
```suggestion
```
that is the default

SuperSandro2000 committed:
Update pkgs/development/python-modules/opensfm/default.nix

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 152957` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>python310Packages.opensfm</li>
<li>python39Packages.opensfm</li>
</ul>
</details>
